File: libraries/python/src/pandaprobe/wrappers/gemini/wrapper.py
"""wrap_gemini — automatic LLM span instrumentation for the Google GenAI client."""

from __future__ import annotations

import functools
import logging
from typing import Any, TypeVar

from pandaprobe.wrappers._base import (
    AsyncStreamReducer,
    SyncStreamReducer,
    close_llm_span,
    error_llm_span,
    safe_serialize,
)
from pandaprobe.wrappers.gemini.utils import convert_config_to_dict, enter_gemini_span

logger = logging.getLogger("pandaprobe")

T = TypeVar("T")


def wrap_gemini(client: T) -> T:
    """Instrument a ``google.genai.Client`` for automatic LLM span tracing.

    Monkey-patches ``models.generate_content``, ``models.generate_content_stream``
    and their async counterparts so every call automatically produces an LLM span.

    Returns the same client instance (mutated).
    """
    _patch_sync_generate(client)
    _patch_sync_stream(client)
    _patch_async_generate(client)
    _patch_async_stream(client)
    return client


# ---------------------------------------------------------------------------
# Patching helpers
# ---------------------------------------------------------------------------


def _patch_sync_generate(client: Any) -> None:
    if not hasattr(client, "models") or not hasattr(client.models, "generate_content"):
        return
    original = client.models.generate_content
    client.models.generate_content = _sync_generate_wrapper(original)


def _patch_sync_stream(client: Any) -> None:
    if not hasattr(client, "models") or not hasattr(client.models, "generate_content_stream"):
        return
    original = client.models.generate_content_stream
    client.models.generate_content_stream = _sync_stream_wrapper(original)


def _patch_async_generate(client: Any) -> None:
    if (
        not hasattr(client, "aio")
        or not hasattr(client.aio, "models")
        or not hasattr(client.aio.models, "generate_content")
    ):
        return
    original = client.aio.models.generate_content
    client.aio.models.generate_content = _async_generate_wrapper(original)


def _patch_async_stream(client: Any) -> None:
    if (
        not hasattr(client, "aio")
        or not hasattr(client.aio, "models")
        or not hasattr(client.aio.models, "generate_content_stream")
    ):
        return
    original = client.aio.models.generate_content_stream
    client.aio.models.generate_content_stream = _async_stream_wrapper(original)


# ---------------------------------------------------------------------------
# Sync wrappers
# ---------------------------------------------------------------------------


def _sync_generate_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = convert_config_to_dict(kwargs)
        return _sync_blocking_generate(original, args, kwargs, cleaned)

    return wrapper


def _sync_stream_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = convert_config_to_dict(kwargs)
        return _sync_streaming_generate(original, args, kwargs, cleaned)

    return wrapper


def _sync_blocking_generate(original, args, kwargs, cleaned):  # noqa: ANN001
    span_ctx = enter_gemini_span(cleaned, "gemini-generate")
    try:
        response = original(*args, **kwargs)
        _finish_gemini_span(span_ctx, response)
        return response
    except Exception as exc:
        error_llm_span(span_ctx, exc)
        raise


def _sync_streaming_generate(original, args, kwargs, cleaned):  # noqa: ANN001
    span_ctx = enter_gemini_span(cleaned, "gemini-generate")
    try:
        stream = original(*args, **kwargs)
        return _GeminiSyncStream(stream, span_ctx)
    except Exception as exc:
        error_llm_span(span_ctx, exc)
        raise


# ---------------------------------------------------------------------------
# Async wrappers
# ---------------------------------------------------------------------------


def _async_generate_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    async def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = convert_config_to_dict(kwargs)
        return await _async_blocking_generate(original, args, kwargs, cleaned)

    return wrapper


def _async_stream_wrapper(original):  # noqa: ANN001
    @functools.wraps(original)
    async def wrapper(*args: Any, **kwargs: Any) -> Any:
        cleaned = convert_config_to_dict(kwargs)
        return await _async_streaming_generate(original, args, kwargs, cleaned)

    return wrapper


async def _async_blocking_generate(original, args, kwargs, cleaned):  # noqa: ANN001
    span_ctx = enter_gemini_span(cleaned, "gemini-generate")
    try:
        response = await original(*args, **kwargs)
        _finish_gemini_span(span_ctx, response)
        return response
    except Exception as exc:
        error_llm_span(span_ctx, exc)
        raise


async def _async_streaming_generate(original, args, kwargs, cleaned):  # noqa: ANN001
    span_ctx = enter_gemini_span(cleaned, "gemini-generate")
    try:
        stream = await original(*args, **kwargs)
        return _GeminiAsyncStream(stream, span_ctx)
    except Exception as exc:
        error_llm_span(span_ctx, exc)
        raise


# ---------------------------------------------------------------------------
# Gemini-specific stream reducers
# ---------------------------------------------------------------------------


class _GeminiSyncStream(SyncStreamReducer):
    def reduce_chunks(self, span_ctx: Any, chunks: list[Any]) -> None:
        _reduce_gemini_stream(span_ctx, chunks)


class _GeminiAsyncStream(AsyncStreamReducer):
    def reduce_chunks(self, span_ctx: Any, chunks: list[Any]) -> None:
        _reduce_gemini_stream(span_ctx, chunks)


def _reduce_gemini_stream(span_ctx: Any, chunks: list[Any]) -> None:
    """Reduce Gemini streaming chunks into final span data."""
    answer_parts: list[str] = []
    thought_parts: list[str] = []
    usage: dict[str, int] | None = None

    for chunk in chunks:
        try:
            candidates = getattr(chunk, "candidates", None)
            if candidates:
                content = getattr(candidates[0], "content", None)
                parts = getattr(content, "parts", None) if content else None
                if parts:
                    for part in parts:
                        text = getattr(part, "text", None)
                        if not text:
                            continue
                        if getattr(part, "thought", False):
                            thought_parts.append(text)
                        else:
                            answer_parts.append(text)
                    continue
            text = getattr(chunk, "text", None)
            if text:
                answer_parts.append(text)
        except Exception:
            pass

    last_chunk = chunks[-1] if chunks else None
    if last_chunk is not None:
        try:
            usage = _extract_gemini_usage(getattr(last_chunk, "usage_metadata", None))
        except Exception:
            pass

    if answer_parts:
        span_ctx.set_output(
            {"messages": [{"role": "assistant", "content": "".join(answer_parts)}]},
        )
    if thought_parts:
        span_ctx.set_metadata({"reasoning_summary": "\n\n".join(thought_parts)})
    if usage:
        span_ctx.set_token_usage(**usage)

    close_llm_span(span_ctx)


# ---------------------------------------------------------------------------
# Response extraction helpers
# ---------------------------------------------------------------------------


def _finish_gemini_span(span_ctx: Any, response: Any) -> None:
    """Extract output, model, reasoning summary, and usage from a Gemini response."""
    if span_ctx is None:
        return

    try:
        answer_texts, thought_texts = _split_parts(response)
        if answer_texts:
            span_ctx.set_output(
                {"messages": [{"role": "assistant", "content": "".join(answer_texts)}]},
            )
        elif not thought_texts:
            text = getattr(response, "text", None)
            if text is not None:
                span_ctx.set_output(
                    {"messages": [{"role": "assistant", "content": text}]},
                )
            else:
                candidates = getattr(response, "candidates", None)
                if candidates:
                    content = getattr(candidates[0], "content", None)
                    if content is not None:
                        serialized = safe_serialize(content)
                        if isinstance(serialized, dict):
                            serialized["role"] = "assistant"
                        span_ctx.set_output({"messages": [serialized]})
        if thought_texts:
            span_ctx.set_metadata({"reasoning_summary": "\n\n".join(thought_texts)})
    except Exception as exc:
        logger.debug("Error extracting Gemini response output: %s", exc)

    try:
        usage = _extract_gemini_usage(getattr(response, "usage_metadata", None))
        if usage:
            span_ctx.set_token_usage(**usage)
    except Exception as exc:
        logger.debug("Error extracting Gemini response usage: %s", exc)

    close_llm_span(span_ctx)


def _split_parts(response: Any) -> tuple[list[str], list[str]]:
    """Separate answer text from thought summaries in a Gemini response.

    Returns ``(answer_texts, thought_texts)``.
    """
    answer_texts: list[str] = []
    thought_texts: list[str] = []

    candidates = getattr(response, "candidates", None)
    if not candidates:
        return answer_texts, thought_texts

    content = getattr(candidates[0], "content", None)
    parts = getattr(content, "parts", None) if content else None
    if not parts:
        return answer_texts, thought_texts

    for part in parts:
        text = getattr(part, "text", None)
        if not text:
            continue
        if getattr(part, "thought", False):
            thought_texts.append(text)
        else:
            answer_texts.append(text)

    return answer_texts, thought_texts


def _extract_gemini_usage(usage_metadata: Any) -> dict[str, int] | None:
    """Map Gemini ``usage_metadata`` fields to our standard token usage dict."""
    if usage_metadata is None:
        return None

    prompt = getattr(usage_metadata, "prompt_token_count", 0) or 0
    completion = getattr(usage_metadata, "candidates_token_count", 0) or 0
    extra: dict[str, int] = {}

    total = getattr(usage_metadata, "total_token_count", 0) or 0
    if total > 0:
        extra["total_tokens"] = total

    thoughts = getattr(usage_metadata, "thoughts_token_count", 0) or 0
    if thoughts > 0:
        extra["reasoning_tokens"] = thoughts

    cached = getattr(usage_metadata, "cached_content_token_count", 0) or 0
    if cached > 0:
        extra["cache_read_tokens"] = cached

    if prompt == 0 and completion == 0 and not extra:
        return None

    return {"prompt_tokens": prompt, "completion_tokens": completion, **extra}

File: libraries/typescript/src/wrappers/gemini/wrapper.ts
/** wrapGemini — automatic LLM span instrumentation for the Google GenAI JS client. */

import type { SpanContext } from "../../tracing/span.js";
import { closeLlmSpan, errorLlmSpan, safeSerialize, tokenUsage, wrapAsyncStream } from "../base.js";
import { enterGeminiSpan } from "./utils.js";

type Any = any;

/**
 * Instrument a `GoogleGenAI` client.
 *
 * Patches `models.generateContent` (blocking) and `models.generateContentStream`
 * (streaming). Returns the same client instance (mutated).
 */
export function wrapGemini<T>(client: T): T {
  patchGenerate(client as Any);
  patchStream(client as Any);
  return client;
}

function patchGenerate(client: Any): void {
  const models = client?.models;
  if (!models || typeof models.generateContent !== "function") {
    return;
  }
  const original = models.generateContent.bind(models);
  models.generateContent = async (params: Any, ...rest: Any[]): Promise<Any> => {
    const span = enterGeminiSpan(params ?? {}, "gemini-generate");
    try {
      const response = await original(params, ...rest);
      finishGeminiSpan(span, response);
      return response;
    } catch (exc) {
      errorLlmSpan(span, exc);
      throw exc;
    }
  };
}

function patchStream(client: Any): void {
  const models = client?.models;
  if (!models || typeof models.generateContentStream !== "function") {
    return;
  }
  const original = models.generateContentStream.bind(models);
  models.generateContentStream = async (params: Any, ...rest: Any[]): Promise<Any> => {
    const span = enterGeminiSpan(params ?? {}, "gemini-generate");
    try {
      const stream = await original(params, ...rest);
      return wrapAsyncStream(stream as AsyncIterable<Any>, span, reduceGeminiStream);
    } catch (exc) {
      errorLlmSpan(span, exc);
      throw exc;
    }
  };
}

function extractGeminiUsage(usageMetadata: Any): Record<string, number> | null {
  if (!usageMetadata) {
    return null;
  }
  const u = usageMetadata;
  return tokenUsage({
    prompt_tokens: u.promptTokenCount ?? u.prompt_token_count,
    completion_tokens: u.candidatesTokenCount ?? u.candidates_token_count,
    total_tokens: u.totalTokenCount ?? u.total_token_count,
    reasoning_tokens: u.thoughtsTokenCount ?? u.thoughts_token_count,
    cache_read_tokens: u.cachedContentTokenCount ?? u.cached_content_token_count,
  });
}

function setUsage(span: SpanContext, usage: Record<string, number> | null): void {
  if (!usage) {
    return;
  }
  const { prompt_tokens, completion_tokens, ...extra } = usage;
  span.setTokenUsage({ promptTokens: prompt_tokens ?? 0, completionTokens: completion_tokens ?? 0, ...extra });
}

function splitParts(response: Any): { answerTexts: string[]; thoughtTexts: string[] } {
  const answerTexts: string[] = [];
  const thoughtTexts: string[] = [];
  const parts = response?.candidates?.[0]?.content?.parts;
  if (!Array.isArray(parts)) {
    return { answerTexts, thoughtTexts };
  }
  for (const part of parts) {
    const text = part?.text;
    if (!text) {
      continue;
    }
    if (part.thought) {
      thoughtTexts.push(text);
    } else {
      answerTexts.push(text);
    }
  }
  return { answerTexts, thoughtTexts };
}

function responseText(response: Any): string | undefined {
  // `.text` is a getter in the JS SDK; guard against throwing accessors.
  try {
    const t = response?.text;
    return typeof t === "string" ? t : undefined;
  } catch {
    return undefined;
  }
}

function reduceGeminiStream(span: SpanContext, chunks: Any[]): void {
  const answerParts: string[] = [];
  const thoughtParts: string[] = [];

  for (const chunk of chunks) {
    const parts = chunk?.candidates?.[0]?.content?.parts;
    if (Array.isArray(parts)) {
      for (const part of parts) {
        const text = part?.text;
        if (!text) {
          continue;
        }
        if (part.thought) {
          thoughtParts.push(text);
        } else {
          answerParts.push(text);
        }
      }
      continue;
    }
    const text = responseText(chunk);
    if (text) {
      answerParts.push(text);
    }
  }

  const last = chunks.length > 0 ? chunks[chunks.length - 1] : undefined;
  const usage = extractGeminiUsage(last?.usageMetadata ?? last?.usage_metadata);

  try {
    if (answerParts.length > 0) {
      span.setOutput({ messages: [{ role: "assistant", content: answerParts.join("") }] });
    }
    if (thoughtParts.length > 0) {
      span.setMetadata({ reasoning_summary: thoughtParts.join("\n\n") });
    }
    setUsage(span, usage);
  } catch {
    // best-effort
  }
  closeLlmSpan(span);
}

function finishGeminiSpan(span: SpanContext | null, response: Any): void {
  if (span === null) {
    return;
  }
  try {
    const { answerTexts, thoughtTexts } = splitParts(response);
    if (answerTexts.length > 0) {
      span.setOutput({ messages: [{ role: "assistant", content: answerTexts.join("") }] });
    } else if (thoughtTexts.length === 0) {
      const text = responseText(response);
      if (text !== undefined) {
        span.setOutput({ messages: [{ role: "assistant", content: text }] });
      } else {
        const content = response?.candidates?.[0]?.content;
        if (content != null) {
          const serialized = safeSerialize(content);
          if (serialized && typeof serialized === "object" && !Array.isArray(serialized)) {
            (serialized as Record<string, unknown>).role = "assistant";
          }
          span.setOutput({ messages: [serialized] });
        }
      }
    }
    if (thoughtTexts.length > 0) {
      span.setMetadata({ reasoning_summary: thoughtTexts.join("\n\n") });
    }
  } catch {
    // ignore
  }

  try {
    setUsage(span, extractGeminiUsage(response?.usageMetadata ?? response?.usage_metadata));
  } catch {
    // ignore
  }

  closeLlmSpan(span);
}
